### PR TITLE
Add out-of-the-box support for Visual Studio Code

### PR DESCRIPTION
**Pull request by Mailaender**
### References
* https://go.microsoft.com/fwlink/?LinkId=827846
* https://go.microsoft.com/fwlink/?LinkId=733558
* https://go.microsoft.com/fwlink/?LinkId=830387

### CODE REVIEW

**Comment by penev92** [comment]
Isn't `.vscode` supposed to be gitignored like any other IDE-specific asset?
Also I don't particularly like the idea of mentioning mono in things that may affect my IDE on Windows. If we swap that for a .NET Core debugger then maybe.

**Comment by teinarss** [comment]
Everything except those files should be added to gitignore,
see attached https://github.com/github/gitignore/blob/master/Global/VisualStudioCode.gitignore

**Comment by phrohdoh** [comment]
IMO https://github.com/OpenRA/OpenRA/commit/2876522617ba83d747f4d5406910cb559244031e sets the precedent of allowing per-editor configuration in-tree.  `.vscode` is "supposed" to be ignored iff that's what the people working on the project want, it isn't a collection of build artifacts.

**Comment by Mailaender** [comment]
This basically adds code completion, enables roslyn analyzers for our code style rules and most importantly adds debug support.
debug
https://marketplace.visualstudio.com/items?itemName=ms-vscode.mono-debug is straight from @Microsoft and should work on Windows, too. It is only the Mono Debugger, because that one is available as Open Source and cross-platform I guess.

**Review by phrohdoh — DISMISSED** [review]

File: .vscode/tasks.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+	"version": "2.0.0",
+	"tasks": [
+		{
+			"label": "build",
+			"command": "msbuild",
+			"args": ["-t:build", "-p:Configuration=Release", "-verbosity:m", "-nologo"]
```
Comment (phrohdoh) [inline .vscode/tasks.json]: The following _may_ be helpful, but I haven't been able to figure out what it really does for us.

```suggestion
			"args": ["-t:build", "-p:Configuration=Release", "-verbosity:m", "-nologo"],
			"problemMatcher": ["$msCompile"]
```

https://code.visualstudio.com/Docs/editor/tasks#_processing-task-output-with-problem-matchers

Comment (Mailaender) [reply .vscode/tasks.json]: It reports compiler errors in the problems pane:

image

However, it is redundant to the Roslyn Analyzers so it duplicates all errors and the paths don't work when click on them. I won't add it until that is fixed.

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+	"version": "0.2.0",
+	"configurations": [
+		{
+			"name": "Launch (TD)",
+			"type": "mono",
```
Comment (obrakmann) [inline .vscode/launch.json]: Apparently there are ways to customize these settings per OS.  Could you please do that to satisfy Penev's concern? After that it should be good to go, a quick test was working quite reliably.

**Review by obrakmann — DISMISSED** [review]

**Comment by deleted-user-1** [comment]
> Everything except those files should be added to gitignore, see attached https://github.com/github/gitignore/blob/master/Global/VisualStudioCode.gitignore

I think it would be enough if we added these to the .gitignore file so that those of us who don't use a proper .NET IDE can exclude their config from version control. I don't know of any maintainer that uses vscode and although it's my favorite editor, I don't like forcing others to maintain my config or advertising something that just is not a good IDE for .NET development (no native .NET Framework support).

**Comment by Mailaender** [comment]
I have actually never encountered those temporary files from the global `.gitignore` although I use it for half a year now.

File: .vscode/launch.json (hunk 2 above)
Comment (Mailaender) [reply .vscode/launch.json]: According to https://github.com/microsoft/vscode-mono-debug#vs-code-mono-debug this should also be correct on Windows. I am not sure what to change here. This is the debugger type, not the runtime launcher.

**Comment by deleted-user-1** [comment]
I meant we don't need this PR if somebody added `/.vscode` to .gitignore (as penev said).

**Comment by Mailaender** [comment]
Figuring this out took me >6 months and I only learned about it by accident. I think we shouldn't ignore the most popular text editor for programmers.

**Comment by deleted-user-1** [comment]
I know, I was there when dragunoff posted their config on discord... It's another reason for not maintaining this in-tree - if the users don't know, why should maintainers who don't even use vscode know or can be expected to do the inevitable adjustments in some years?

Couldn't we just add a wiki page and let that rotten like all the others?

**Comment by Mailaender** [comment]
[88 k](https://github.com/search?l=JSON&q=%22cwd%22%3A+%22%24%7BworkspaceRoot%7D%22%2C&type=Code) people disagree with you.

File: .vscode/launch.json (hunk 2 above)
Comment (obrakmann) [reply .vscode/launch.json]: Can we get a confirmation from a Windows dev for that then, please? @OpenRA/bug-assignees

**Review by deleted-user-1 — COMMENTED** [review]
FWIW, I'd be glad about getting some more functionality out of vscode but dislike the ."vscode" foldername- advertisement. The description for recommended extensions is  *This extension is recommended by users of the current workspace.* fyi, so not sounding official, which eliminates some of my concerns.

Unfortunately, microsoft/vscode-mono-debug@21a0b1433e8ca281fd1f9fc15ba395ef33ed2686 (for latest version 0.15.7 however in vscode, the version is 0.15.8) looks a bit abandoned. Only non-conditional breakpoints will work to inspect local variables - the other two types of breakpoints (Log, Expression) are not supported (hover over the breakpoint shortly after starting the f5-session). This worked for me rather stable on code -v
1.46.1
cd9ea6488829f560dc949a8b2fb789f3cdc05f5d
x64

**Comment by Mailaender** [comment]
The `.vscode` folder name is hard-coded and required. It is invisible by default on Linux if that helps.

**Comment by deleted-user-1** [comment]
The fact that you can't change it does not invalidate anything I said in my review and that you can hide it in your filesystem does not change that we'll be advertising vscode on github or in the source generally. What's your comment on the situation with the abandoned extension?

**Comment by Mailaender** [comment]
It is true that this is not being developed any further https://github.com/microsoft/vscode-mono-debug/issues/50 However there is already .NET Core https://github.com/OpenRA/OpenRA/issues/15954 on the horizon.

**Comment by deleted-user-1** [comment]
> It is true that this is not being developed any further [microsoft/vscode-mono-debug#50](https://github.com/microsoft/vscode-mono-debug/issues/50) However there is already .NET Core #15954 on the horizon.

Thanks, I hadn't noticed that there's an issue for it.

**Comment by SteelPhase** [comment]
Including `.vscode` subdirectory is a best practice for Visual Studio Code. The gitignore file listed above includes all the files that should be included from `.vscode`. I wouldn't consider it an advertisement of which IDE to use, but it makes your repository easier to pick up in that editor. Changes are relatively infrequent to that folder, and it being stale shouldn't be an issue. Just thought I'd throw in some info about this. Pretty much use vscode as my daily driver.

File: .vscode/launch.json (hunk 2 above)
Comment (Mailaender) [reply .vscode/launch.json]: @dragunoff initially suggested those and I believe he was running Windows.

Comment (dragunoff) [reply .vscode/launch.json]: The config I posted was for Linux. I also run Windows but have not utilized vscode for debugging there. Trying to use the config in this PR on Win7 gives me this error:
**изображение**

The debugger type for Windows should be `coreclr` but even then it doesn't attach to the process properly. So the configuration in this PR is only working on Linux. I have not dug into vscode launch configs but I guess there is no good way to support both platforms. That being said I think it's best to not include `.vscode` in the repo but definitely document these configurations somewhere (in a separate repo or in a wiki).

Comment (pchote) [reply .vscode/launch.json]: Does vscode support splitting config files to allow the common config to exist in the repo, and then local overrides to set the debugger without modifying tracked files?

Comment (dragunoff) [reply .vscode/launch.json]: Turns out it does support platform specific overrides: https://code.visualstudio.com/Docs/editor/debugging#_platformspecific-properties

Comment (pchote) [reply .vscode/launch.json]: Assuming that the current setup works on macOS (I will need to test this) then a windows-specific override can set whatever is needed to get it running there.

Comment (Mailaender) [reply .vscode/launch.json]: @dragunoff You might need https://www.mono-project.com/download/stable/#download-win plus the extension. As it says debugger not supported my guess is you didn't install the extension. `coreclr` needs OpenRA to build using .NET Core which is https://github.com/OpenRA/OpenRA/pull/17989.

Comment (pchote) [reply .vscode/launch.json]: When does the error dialog above display? When opening the project, when trying to launch it, or when trying to debug it?

Comment (dragunoff) [reply .vscode/launch.json]: It displays when attempting to launch the project from the debug panel in vscode.

Comment (pchote) [reply .vscode/launch.json]: I find it crazy that Microsoft would force you to run .NET applications under mono on windows.

https://stackoverflow.com/questions/47707095/visual-studio-code-for-net-framework suggests that `"type": "clr"` should work.

Comment (Mailaender) [reply .vscode/launch.json]: That might disable the debugging feature.

Comment (Mailaender) [reply .vscode/launch.json]: Actually it shouldn't. Updated for Windows.

**Comment by penev92** [comment]
When I wrote my initial comment I was thinking these are IDE junk files the likes of which should be gitignored by default.
But having IDE configuration files in the repository is not uncommon in one form or another, so these specific ones being there isn't harmful nor a precedent.
As for the funtionality, we established [on Discord](https://discordapp.com/channels/153649279762694144/160191413256978433/730519996668837991) that this works on Windows after installing the recommended extensions **without** requiring Mono to be installed (which would have been a 👎 from me).
Since this doesn't affect anything on Windows (you can't debug with VSCode and .NET Framework anyway) and enables debugging on other operating systems (or on Windows, if you install Mono ... presumably) and committing IDE-specific configurations isn't really a precedent, I think we can safely merge this with one requested change:
> @cjshmyr 
perhaps the default would want to be clr with osx/linux overrides, assuming they require that extension
in the sense that it adds extra features to vscode, but it's not installing mono
but yeah I feel the default is appropriately "clr", not "mono", since the mono debug extension isn't required for windows -- and the OS specific override json isn't respected unless mono debug is installed anyways

On a side note once the move to .NET Core (or .NET 5?) is done, there will also be no more issue with debugging on Windows.

**Comment by SteelPhase** [comment]
Just tested on windows. Had to remove the brackets around `"clr"`

```json
"windows": {
    "type": "clr"
},
```

**Comment by penev92** [comment]
Oh yeah also that, forgot to mention, but @cjshmyr also told me to do that when I was testing so it can run.

**Comment by Mailaender** [comment]
Updated. Sorry, I was copy pasting the `args` example which had an array as argument.

**Comment by cjshmyr** [comment]
I think my comment got missed, but what do you think about making the default "clr", and setting linux and osx overrides to use "mono"? Essentially reverse of the current setup.

The operating system overrides in `launch.json` won't be recognized unless you install the `Mono Debug` vscode extension, which does not serve a purpose for Windows / CLR. My assumption is Linux or OSX users would be installing this extension.

**Comment by Mailaender** [comment]
Makes sense. Updated.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
+{
+	"version": "0.2.0",
+	"configurations": [
+		{
+			"name": "Launch (TD)",
+			"type": "clr",
+			"linux": {
```
Comment (cjshmyr) [inline .vscode/launch.json:7]: I think we have to add "osx" specific overrides too
```
"osx": {
	"type": "mono"
},
```

Comment (pchote) [reply .vscode/launch.json:7]: Yes, confirmed that this is needed and works (to a limited degree: all mods NRE in `DropDownButtonWidget` when I try to run them on this branch)

Comment (Mailaender) [reply .vscode/launch.json:7]: Okay ~~, untested though~~.

Comment (pchote) [reply .vscode/launch.json:7]: Ok, it looks like the debug task just doesn't force a compile first... weird choice there vscode.

**Review by pchote — COMMENTED** [review]
Verified that the out-of-box experience + building + debugging seems to work on macOS, although the vscode workflow is a bit awkward.

**Comment by pchote** [comment]
Verified that the updated version still works on macOS, and also works on Ubuntu 20.04.

Can we get confirmation from our windows users? Then this should hopefully be good to go

IMO it would be good to squash all the commits down to one though, theres no need to record the workarounds into the git history.

**Review by cjshmyr — DISMISSED** [review]
Works on Windows. Tested building (`ctrl-shift-b` -> build) from a clean environment.

Also please squash the commits!

**Comment by Mailaender** [comment]
Squashed it into one.

**Comment by penev92** [comment]
Can confirm this now makes VSCode actually usable, and has addressed all the concerns discussed across mediums.
On Windows you can't debug unless you install Mono, but that has always been the case and nobody expects otherwise.

My impression from last night:
> this actually got a lot more helpful/usable than I expected
 the only difference between VS and VSCode atm is I need to press an extra Enter (or 2) in VSCode to build and I get no working breakpoints (because I didn't install Mono)
soooo it's almost perfect for me :ok_hand:

So 👍

**Review by penev92 — APPROVED** [review]
👍

**Pull request merged**